### PR TITLE
[SWY-101] Add resource preview metadata

### PR DESCRIPTION
**Pull request by justaddcl**
## Summary
[SWY-101]
- add server-side resource metadata fetching with safe URL validation, redirect handling, HTML extraction, and failed-preview persistence
- make resource names optional in the form while keeping the database `resources.title` column non-null for existing environments
- use title/meta title/hostname fallback display, add create/edit drawer preview cards, auto-suggest blank names from fetched page titles, and add a resource-card refresh preview action

|    | Mobile | Desktop |
| - | ------ | -------- |
| Link a resource preview | SWY-101__after--mobile--link-resource | SWY-101__after--desktop--link-resource | 
| Refresh resource preview | SWY-101__after--mobile--refresh-preview | SWY-101__after--desktop--refresh-preview |
| Edit resource preview | SWY-101__after--mobile--edit-resource | SWY-101__after--desktop--edit-resource |

## Validation
- `pnpm lint` (passes; repo has existing warnings)
- `pnpm jest src/server/orpc/services/resource/__tests__/resourceMetadata.test.ts src/server/orpc/services/resource/__tests__/updateResource.test.ts src/server/orpc/services/resource/__tests__/deleteResource.test.ts src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx src/modules/songs/components/ResourceCard/__tests__/ResourceCard.test.tsx --runInBand`

## Notes
- Blank resource names are stored as an empty string so saves work against the current non-null `resources.title` constraint; display logic treats blank titles as absent and falls through to preview metadata or hostname.
- The metadata fetch cap is 2 MiB so pages with larger document heads, including YouTube watch pages, can still expose OpenGraph tags before the request is capped.
- The focused component suite still logs an existing Next/Image `fetchPriority` warning from the empty-state artwork.

## Summary by CodeRabbit

* **New Features**
  * Rich resource previews with automatic metadata (image, title, description) and a visual preview panel.
  * "Refresh preview" action to update a resource’s preview on demand.
  * Improved title suggestion that auto-fills from preview metadata while creating a resource.

* **Bug Fixes**
  * More robust URL normalization and previewability checks.
  * Improved trimming/handling of resource titles and consistent display titles across UI.

[Review Change Stack](https://app.coderabbit.ai/change-stack/justaddcl/sanbi/pull/76?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<h3>Greptile Summary</h3>

This PR adds server-side resource preview metadata fetching with SSRF protection, redirect following, HTML metadata extraction, and a "Refresh preview" action. It also makes the resource name field optional, wires up a preview panel in create/edit drawers, and auto-suggests titles from fetched page metadata.

- **`resourceMetadata.ts`** implements SSRF-safe HTTPS fetching with DNS pinning, per-request timeout, a 2 MiB byte cap that resolves (not rejects) buffered HTML, and a `fetchResourcePreviewMetadata` helper that falls back to `status: "failed"` on any error.
- **`refreshResourceMetadata.ts`** and the router wiring expose a `POST /refresh-metadata` endpoint that throws `BAD_REQUEST` on fetch failure, preserving existing metadata instead of clearing it.
- **`resourceTitleSchema`** replaces the former required-title validator; empty/whitespace titles transform to `null`, the DB schema is correspondingly made nullable, and `getResourceDisplayTitle` falls through to `metaTitle` then hostname for display.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The SSRF protection, byte-cap fix, entity-decoding fix, and refresh-clears-metadata fix from prior review rounds are all correctly implemented in the current HEAD.

All previously identified blocking issues have been addressed: the response byte cap now resolves buffered HTML instead of rejecting it, sanitizeNullableMetadataText correctly double-decodes to avoid entity re-encoding, and refreshResourceMetadata throws before writing when the fetch fails. The remaining observations are quality suggestions without production impact.

No files require special attention. The two suggestions (cumulative redirect timeout in resourceMetadata.ts and .localhost guard in resourcePreviewUrl.ts) are low-risk improvements that can be addressed in a follow-up.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/server/orpc/services/resource/resourceMetadata.ts | New SSRF-safe metadata fetching service with DNS pinning, redirect following, byte capping (resolves buffered HTML), and graceful status:failed fallback. Well-structured and thoroughly tested. |
| src/server/orpc/services/resource/refreshResourceMetadata.ts | New service for refreshing resource metadata; correctly throws BAD_REQUEST on fetch failure instead of wiping existing metadata. Organization ownership check is clean. |
| src/server/utils/htmlMetadata/htmlMetadata.ts | New HTML metadata extractor using regex + he + DOMPurify. Double-decode pattern in sanitizeNullableMetadataText correctly prevents entity re-encoding. resolveMetadataUrl handles relative URLs safely via validateUrl. |
| src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx | Adds debounced URL preview query with staleTime and refetchOnWindowFocus:false (addressing prior feedback). Auto-title suggestion logic with autoGeneratedTitle state is complex but well-guarded via two coordinated effects. |
| src/lib/types/zod.ts | New resourceTitleSchema makes title optional/nullable and transforms empty strings to null. Length refine runs on the sanitized value which is appropriate. New output and input schemas for preview metadata are well-typed. |
| src/server/orpc/routers/resource.ts | Adds previewMetadata (POST, read-only with URL in body) and refreshMetadata endpoints with proper org-membership authorization checks. createResource now inlines metadata fetch via resolveResourceMetadataForUrl. |
| src/server/db/schema.ts | resources.title changed from .notNull() to nullable to support blank resource names. No migration file (project uses drizzle-kit push); operational db:push:prod must be run against each environment before deploying. |
| src/lib/urls/resourcePreviewUrl.ts | New client-side URL utilities for determining previewability and extracting display hostname. isPreviewableResourceUrl correctly rejects IP literals, non-HTTPS, and dotless hostnames; server-side SSRF checks cover residual edge cases. |
| src/server/orpc/services/resource/updateResource.ts | URL changes now trigger resolveResourceMetadataForUrl, atomically updating preview columns alongside the URL. Title handling correctly accepts null to clear the field. isUniqueConstraintViolation moved to shared postgres.ts. |

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
src/server/orpc/services/resource/resourceMetadata.ts:231-260
**No cumulative timeout across the redirect chain**

Each request in `fetchHtmlWithRedirects` gets a fresh `timeoutMs` (default 5 s). With `MAX_REDIRECTS = 5`, a pathological chain of slow-but-not-timing-out responses can hold a server thread for up to `6 × 5 s = 30 s`. Because `resolveResourceMetadataForUrl` is called inline inside the `createResource` and `updateResource` route handlers, this directly blocks those user-facing mutations. Consider tracking an overall deadline — e.g., record `Date.now()` before the loop and pass a shrinking `timeoutMs` (or an `AbortSignal`) to each iteration so the total wall time is bounded.

### Issue 2 of 2
src/lib/urls/resourcePreviewUrl.ts:9-13
**`.localhost` TLD subdomains pass the client-side previewability check**

`isPreviewableResourceUrl("https://evil.localhost/path")` returns `true` because `evil.localhost` contains a dot and isn't an IP literal. The query is enabled and a POST to `/preview-metadata` fires, which then fails the server-side SSRF check in `isBlockedHostname`. Adding a `.localhost` guard here mirrors the server-side check and avoids the unnecessary round-trip.

```suggestion
    return (
      parsedUrl.protocol === "https:" &&
      hostname.includes(".") &&
      !isIpLiteralHostname(hostname) &&
      hostname !== "localhost" &&
      !hostname.endsWith(".localhost")
    );
```

`````

</details>

<sub>Reviews (8): Last reviewed commit: ["Restore nullable resource title behavior"](https://github.com/justaddcl/sanbi/commit/877b53afe45838e654bd59f00c5cb72ec8be5bc7) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=33130168)</sub>

### CODE REVIEW

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 1]
```diff
@@ -29,12 +34,102 @@ const resourceFormSchema = insertResourceSchema.pick({
...
+    )
+    ?.trim();
+
+const ResourceFormPreview: React.FC<{
+  resource?: Resource;
+  title: string | null | undefined;
+  url: string;
+  previewMetadata?: ResourcePreviewMetadata;
+  isLoading: boolean;
+}> = ({ resource, title, url, previewMetadata, isLoading }) => {
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx]: Addressed in 0c56e11 by extracting ResourceFormPreviewProps and switching ResourceFormPreview to a typed named arrow function.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 2]
```diff
@@ -224,10 +348,15 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
...
-            url: watchedUrl ?? "",
-          })}
+          <ResourceFormPreview
+            resource={resource}
+            title={watchedTitle ?? ""}
+            url={watchedUrl ?? ""}
+            previewMetadata={previewMetadataQuery.data}
+            isLoading={
+              isWaitingForPreviewDebounce || previewMetadataQuery.isFetching
+            }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx:355]: Addressed in 0c56e11 by only passing preview metadata when the current URL is previewable and matches the settled debounced URL. This avoids stale metadata on non-previewable URLs while preserving metadata for valid redirect targets.

File: src/server/orpc/routers/resource.ts [hunk 3]
```diff
@@ -8,16 +8,31 @@ import {
...
+  POSTGRES_UNIQUE_CONSTRAINT_VIOLATION_CODE,
+  updateResourceForOrganization,
+} from "@server/orpc/services/resource/updateResource";
 
 const ROUTER_PREFIX = "/resource";
 
+const isUniqueConstraintViolation = (error: unknown) =>
+  error instanceof Error &&
+  "code" in error &&
+  error.code === POSTGRES_UNIQUE_CONSTRAINT_VIOLATION_CODE;
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/routers/resource.ts]: Addressed in 0c56e11 by moving the Postgres unique-constraint code and predicate into src/server/utils/db/postgres.ts and importing the shared helper from both call sites.

File: src/server/orpc/services/resource/resourceMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+      rawTitle ? decodeHtmlEntities(rawTitle) : null,
+      300,
+    ),
+    description: sanitizeNullableMetadataText(
+      rawDescription ? decodeHtmlEntities(rawDescription) : null,
+      500,
+    ),
+    imageUrl: resolveMetadataUrl(rawImage, normalizedUrl),
+    faviconUrl: resolveMetadataUrl(rawFavicon, normalizedUrl),
+  };
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/resourceMetadata.ts]: Addressed in 9dbaf1a by removing the shared sanitizeInput helper from metadata text sanitation. Metadata text now decodes entities, strips tags with DOMPurify only, decodes again for plain-text storage, and has a regression assertion for `Hello &amp; World` -> `Hello & World`.

File: src/server/orpc/services/resource/resourceMetadata.ts [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+              totalBytes += chunk.byteLength;
+
+              if (totalBytes > maxHtmlBytes) {
+                request.destroy(
+                  new Error(
+                    "Resource metadata response exceeded the size limit",
+                  ),
+                );
+                return;
+              }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/resourceMetadata.ts:211]: Addressed in 97663d9. The response size cap now resolves with the buffered HTML already collected and then destroys the request without an error, so pages that exceed the cap can still produce metadata from the fetched head/body prefix instead of falling through to `status: "failed"`.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 6]
```diff
@@ -106,6 +110,102 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
...
+      input: {
+        organizationId: requestedOrganizationId ?? "",
+        url: trimmedDebouncedUrl || "https://example.com",
+      },
+    }),
+    enabled:
+      !!requestedOrganizationId &&
+      canPreviewUrl &&
+      !isWaitingForPreviewDebounce,
+  });
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx:140]: Addressed in f464705 by giving the preview metadata query a 5-minute staleTime and disabling refetchOnWindowFocus, so focusing the tab no longer refetches and temporarily disables submit.

File: src/lib/urls/resourcePreviewUrl.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+export const isPreviewableResourceUrl = (url: string) => {
+  try {
+    const parsedUrl = new URL(url);
+
+    return parsedUrl.protocol === "https:" && parsedUrl.hostname.includes(".");
+  } catch {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/urls/resourcePreviewUrl.ts:16]: Addressed in f464705 by rejecting IP-literal hostnames in isPreviewableResourceUrl and adding regression coverage for IPv4 and IPv6 literal URLs.

File: src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx [hunk 8]
```diff
@@ -302,7 +320,7 @@ describe("SongResources resource editing", () => {
       expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
     });
     expect(
-      screen.getByRole("link", { name: new RegExp(resource.title, "i") }),
+      screen.getByRole("link", { name: new RegExp(resourceDisplayTitle, "i") }),
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx]: Addressed in f464705 by replacing the dynamic RegExp accessible-name lookup with a non-regex matcher that treats the generated title as literal text.

File: src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx [hunk 9]
```diff
@@ -364,20 +382,82 @@ describe("SongResources resource editing", () => {
...
+    await openCreateDrawer();
+
+    fireEvent.change(screen.getByLabelText("URL *"), {
+      target: { value: newResourceUrl },
+    });
+
+    await waitFor(() => {
+      expect(screen.getByLabelText("Name")).toHaveValue(suggestedResourceName);
+    });
+    expect(screen.getByLabelText("Name")).toHaveClass("text-slate-500");
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx]: Addressed in f464705 by removing the Tailwind class assertions from the workflow test and keeping the user-visible value and hint assertions.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 10]
```diff
@@ -177,13 +277,23 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
...
         await createResourceMutation.mutateAsync({
           songId,
           organizationId,
           title: formValues.title,
           url: formValues.url,
+          previewMetadata:
+            previewMetadataQuery.data &&
+            formValues.url.trim() === trimmedDebouncedUrl
+              ? previewMetadataQuery.data
+              : undefined,
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx]: Addressed in f464705 by using previewMetadataForCurrentUrl for both create and update submit payloads, instead of sending raw previewMetadataQuery.data.

File: src/modules/songs/utils/getResourceDisplayTitle.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+  if (metaTitle) {
+    return metaTitle;
+  }
+
+  try {
+    return new URL(resource.url).hostname.replace(/^www\./, "");
+  } catch {
+    return resource.url;
+  }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/utils/getResourceDisplayTitle.ts]: Addressed in f464705 by delegating getResourceDisplayTitle hostname fallback to the shared getHostnameFromUrl helper.

File: src/server/orpc/routers/resource.ts [hunk 12]
```diff
@@ -239,12 +279,98 @@ export const deleteResource = organizationProcedure
...
 export const resourceRouter = authedProcedure.prefix(ROUTER_PREFIX).router({
   // QUERIES
   getBySongId: getResourcesBySongId,
+  previewMetadata: previewResourceMetadata,
 
   // MUTATIONS
   create: createResource,
   update: updateResource,
   delete: deleteResource,
+  refreshMetadata: refreshResourceMetadata,
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/routers/resource.ts:332]: Addressed in f464705 by documenting that previewMetadata is a read-only endpoint that uses POST so clients can send the URL in the body.

File: src/server/orpc/services/resource/__tests__/updateResource.test.ts [hunk 13]
```diff
@@ -27,6 +25,19 @@ jest.mock("@orpc/client", () => {
...
+    metadataValues: {
+      status: "ready",
+      metaTitle: "Fetched resource title",
+      metaDescription: null,
+      faviconUrl: null,
+      imageUrl: null,
+      lastFetchedAt: new Date("2026-01-01T00:00:00Z"),
+    },
+  })),
+}));
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/__tests__/updateResource.test.ts]: Addressed in f464705 by making resolveResourceMetadataForUrl an explicit per-test mock and asserting no resolver calls on paths that should not fetch metadata.

File: src/server/orpc/services/resource/resourceMetadata.ts [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+              resolveBufferedResponse();
+            });
+          },
+        );
+
+        request.on("timeout", () => {
+          request.destroy(new Error("Resource metadata request timed out"));
+        });
+        request.on("error", reject);
+        request.end();
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/resourceMetadata.ts:227]: Addressed in f464705 by adding a response stream error handler that rejects through the same settled-response guard as the other stream paths.

File: src/server/orpc/services/resource/updateResource.ts [hunk 15]
```diff
@@ -74,24 +74,22 @@ export const updateResourceForOrganization = async ({
...
-      throw new ORPCError("BAD_REQUEST", {
-        message: "Resource title cannot be empty",
-      });
-    }
-
-    if (title !== resourceToUpdate.title) {
-      updateValues.title = title;
+    if (nextTitle !== resourceToUpdate.title) {
+      updateValues.title = nextTitle;
     }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/updateResource.ts:82]: Addressed in f464705 by storing the trimmed title value and adding test coverage for whitespace-padded titles.

File: src/server/orpc/routers/resource.ts [hunk 16]
```diff
@@ -239,12 +279,99 @@ export const deleteResource = organizationProcedure
...
+        resourceToRefresh.url,
+      );
+      [updatedResource] = await db
+        .update(resources)
+        .set({
+          url: refreshedMetadata.normalizedUrl,
+          ...toResourceMetadataWriteValues(refreshedMetadata),
+        })
+        .where(eq(resources.id, input.resourceId))
+        .returning();
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/routers/resource.ts]: Addressed in 9969e66. Refresh metadata now goes through refreshResourceMetadataForOrganization, which throws a BAD_REQUEST before any database update when fetchResourcePreviewMetadata returns status failed. I also added service coverage proving failed refreshes do not call updateResourceMetadata, so existing preview fields are preserved instead of being cleared.

File: src/lib/types/zod.ts [hunk 17]
```diff
@@ -309,16 +309,38 @@ const resourceUrlSchema = z
...
 
+const resourceTitleSchema = z
+  .string()
+  .transform((title) => {
+    const sanitizedTitle = sanitizeInput(title).trim();
+
+    return sanitizedTitle.length > 0 ? sanitizedTitle : null;
+  })
+  .optional()
+  .nullable();
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/types/zod.ts:322]: Addressed in https://github.com/justaddcl/sanbi/commit/3e0f6aa: `resourceTitleSchema` now sanitizes/trims before enforcing the 255-character DB limit, so overlong titles fail validation before write.

File: src/modules/songs/forms/CreateResourceForm/__tests__/ResourceFormPreviewThumbnail.test.tsx [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+  it("renders the preview image when an image URL is available", () => {
+    const imageUrl = "https://example.com/preview.jpg";
+    const { container } = render(
+      <ResourceFormPreviewThumbnail imageUrl={imageUrl} />,
+    );
+
+    expect(container.querySelector("img")).toHaveAttribute("src", imageUrl);
+    expect(container.querySelector("svg")).not.toBeInTheDocument();
+  });
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/__tests__/ResourceFormPreviewThumbnail.test.tsx]: Addressed in https://github.com/justaddcl/sanbi/commit/3e0f6aa: the thumbnail tests now use Testing Library role queries for the rendered decorative image and no longer assert raw SVG internals.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 19]
```diff
@@ -106,6 +111,104 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
...
+      title: suggestedTitleForCurrentUrl,
+      url: trimmedDebouncedUrl,
+    });
+  }, [
+    autoGeneratedTitle,
+    resourceForm,
+    suggestedTitleForCurrentUrl,
+    trimmedDebouncedUrl,
+    watchedTitle,
+  ]);
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx:227]: Addressed in https://github.com/justaddcl/sanbi/commit/3e0f6aa: the form now clears a generated title when the URL changes to a preview with no title, and there is regression coverage for that workflow.

File: src/server/orpc/services/resource/__tests__/resourceMetadata.test.ts [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+    constructor(code: string, options: { message?: string; cause?: unknown }) {
+      super(options.message);
+      this.name = "ORPCError";
+      this.code = code;
+      this.cause = options.cause;
+    }
+  }
+
+  return { __esModule: true, ORPCError };
+});
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/orpc/services/resource/__tests__/resourceMetadata.test.ts:16]: Addressed in https://github.com/justaddcl/sanbi/commit/3e0f6aa: `resourceMetadata.test.ts` now reuses `mockOrpcErrorModule` from `@testUtils/orpc/mockOrpcError`.

File: src/server/db/schema.ts [hunk 21]
```diff
@@ -293,7 +293,7 @@ export const resources = createTable(
       .references(() => organizations.id)
       .notNull(),
     url: text("url").notNull(),
-    title: varchar("title", { length: 255 }).notNull(),
+    title: varchar("title", { length: 255 }),
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/db/schema.ts:296]: Addressed in https://github.com/justaddcl/sanbi/commit/3e0f6aa by avoiding the nullable-column change: `resources.title` remains `.notNull()`, blank manual titles are stored as an empty string, and display fallback still uses metadata/hostname when the title is blank.

[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/db/schema.ts:296]: Correction after follow-up: the migration exists, so nullable titles are intentional. Follow-up commits https://github.com/justaddcl/sanbi/commit/ad4171d and https://github.com/justaddcl/sanbi/commit/877b53a restore the nullable schema and `null` persistence for blank titles while keeping the sanitized 255-character validation.
